### DISCUSSION TITLE
2024-07-19 금요일

### DISCUSSION THREAD

**Discussion by Captain-Kim** [nbc-final-a10/TripBuddy#60] [기록보드]
2024-07-19 금요일 기록보드입니다.

## 여행가고십조(A10) 일일 공통 일과표

| 시간 | 내용 | 기록 |
| :-: | :-: | :-: |
| 09:00 ~ 10:00 | 개발자 회의 (의사코드 공유) | |
| 10:00 ~ 13:00 | 오전 디자이너 미팅 (와이어프레임 확정) | |
| 13:00 ~ 14:00 | 점심 식사 | |
| 13:00 ~ 16:00 | 오후 디자이너 미팅 (와이어프레임 확정) | |
| 16:00 ~ 18:00 | 의사코드 공유 + DB 테이블 설계 | |
| 18:00 ~ 19:00 | 저녁 식사 | |
| 19:00 ~ 20:00 | 의사코드 공유 + DB 테이블 설계 + 페이지 및 컴포넌트 분배 확정 | |
| 20:00 ~ 21:00 | 회고 | |

### 오늘 한 일

- 본인인증 외부 API 응답값 확인 및 테스트
```
// 사진
<img src="이미지 주소" weight="300" height="300"/>
```

### 내일 계획

- supabase auth 비밀번호 찾기 기능 연결

**Comment by Captain-Kim** [comment]
## 2024-07-19 금요일 기록보드 (김병준)

### 오늘 한 일

- 디자이너 미팅 : 와이어프레임 확정 (100%)
- 개인 의사 코드 공유(하기)
- DB 테이블 셋업 (100%)
- API 명세서 작성 (90%, 10%는 작업 하면서 채워질 것으로 예상)

### 다음 주 월요일 할 일

## 의사코드

### 여정(파티) 만들기

- 글쓰기 카테고리 설정 컴포넌트 :: '버디즈 모집을 시작해 볼까요?'
  - seq 5개 이동 방식, 다음 버튼 클릭시 다음 시퀀스로 이동하는 방식
  - 1) 버디즈 모집을 시작해 볼까요?
    - 몇 명의 버디를 모집할까요?
      - 대표 이미지 1장 선택 -> 5mb 이상 이미지 선택 불가 유효성 검사 -> 미리보기 렌더링 (fileReader, state) -> 글 작성을 눌러야 webp 변환 후 버킷 전송 및 getPublicUrl 후 trips 테이블 trip_image에 쓰기
      - 본인 제외 최대 3명 선택 가능한 버튼 제공 '1명', '2명', '3명' -> 인원 수 클릭 시 `setWantedBuddies` -> 미선택, 복수선택 불가능 유효성 검사 -> '다음' 버튼 클릭 시 다음 시퀀스로 이동
    - 어떤 유형의 여행을 원하세요?
      - chips(해시태그)를 최대 8개 버튼 제공 (chips의 구체적인 내용은 논의 필요) -> 미선택, 3개 초과 선택 불가능 유효성 검사 -> 3개 클릭 시 해당 버튼의 내용을 string으로 `setThemes` -> '다음'버튼 클릭 시 다음 시퀀스로 이동
    - 원하는 버디즈의 특성을 알려주세요.
      - 성별: 여성, 남성, 상관 없음 -> 미선택, 1개 초과 선택 불가능 유효성 검사 -> 여성(female), 남성(male), 상관없음(none) 으로 `setWantedSex` -> 나이는 빼는 것이 좋을 것 같음(실제 사이트 Ref 찾을 수 없음, 유저가 입력한 온보딩에서 입력한 나이의 진위여부도 확인하기 어려움) -> 

```tsx
// 미리보기
"use state";

import { useState } from 'react';
import { createClient } from '@supabase/supabase-js';

const supabaseUrl = '';
const supabaseKey = '';
const supabase = createClient(supabaseUrl, supabaseKey);

function App() {
  const [file, setFile] = useState(null);
  const [previewUrl, setPreviewUrl] = useState(null);
  const [statusMessage, setStatusMessage] = useState('');

  const handleFileChange = (event) => {
    const selectedFile = event.target.files[0];
    if (selectedFile) {
      setFile(selectedFile);
      const reader = new FileReader();
      reader.onload = (e) => {
        setPreviewUrl(e.target.result);
      };
      reader.readAsDataURL(selectedFile);
    }
  };

  const handleUpload = async () => {
    if (file) {
      setStatusMessage('Uploading...');
      try {
        const { data, error } = await supabase.storage
          .from('버킷')
          .upload(`public/${file.name}`, file);
        if (error) {
          throw new Error('업로드 실패');
        }
        setStatusMessage('업로드 성공');
      } catch (error) {
        setStatusMessage('업로드 실패');
      }
    }
  };

  return (
    <div>
      <input type="file" accept="image/*" onChange={handleFileChange} />
      {previewUrl && <img src={previewUrl} alt="Image Preview" style={{ maxWidth: '100%', height: 'auto' }} />}
      {file && <button onClick={handleUpload}>Upload</button>}
      <p>{statusMessage}</p>
    </div>
  );
}

export default App;
```

### 여정(파티) 수정하기

### 버디즈 랭킹 리스트

### 마이페이지(프로필)

- 기존 trip detail page에서 buddy_master의 온도지수가 렌더링되어 있지 않음. -> 개발자 논의 후 디자이너님께 전달

- 온도 측정 로직
  - 온도 단위: int? num?
  - 온도 dafault : 36.5? 36? 0? => 0부터 시작하는 게 계산하기 편할 것 같은데 36.5도로 계산하면... default 온도를 전역 상태로 관리해야 할 듯
  - 온도 측정 로직 : 증가 및 감소 요인
    - 팔로워 1명 증가/감소(단 같은 사람이 팔로우-취소 반복 매크로 작업 식별해야 함)
    - 온보딩 완료 시
    - 결제 여부(isPro) : 결제 1회당?
    - trip 개설, 참여 1회 당?? (마찬가지로 매크로 식별요)
    - 스토리 작성 1회당 ?? (매크로 식별요)
    - 채팅 회당??
    - 차감 로직이 복잡할 것 같음. 그러나 차감로직이 없다면 실제 프로덕션 레벨에서는 인플레이션 발생 가능.
- 만약 이 로직들이 모두 구현된다면 온도지수 테이블을 별도로 만들어야 할 것 같고, 글쓰기 발생 시 테이블에 temp_id, buddy_id, created_at, event, count (0.1, 1, 5... 증가된 온도지수 입력) -> 전부 더해서 온도지수 업데이트 (그런데 언제 업데이트 할 것인가? SQL로 일마다 업데이트해서 buddies 테이블의 buddy_temperature 컬럼에 덮어쓰기?, 로그인 시점에 update? -> 미접속 1일 discount 등의 로직을 구현하긴 힘들듯)
- 당근마켓 Ref라면 Chart.js?? 등 필오할 듯.
- 온도 감소는 미접속 패널티만 주는 것이 맞을 것 같고, 증가 로직은 유틸 함수 제작 후 각 로직 담당자가 추가해주어야 할 듯. -> 아니면 sql 짜야 함(저 많은 테이블의 state change를 tracking 한다면 서버컴 리소스 낭비 우려??)

**Comment by eunohhh** [comment]
### 스토리 의사코드 작성해보기

- / 에서, me가 있으면..  로그인 상태면 + 버튼이 보일 것이고, 이걸 누르면
- /addStory 로 이동하거나, 전체화면 모달을 띄우거나(이동이 나아보임) *의사결정 필요
- 사진첩에서 선택의경우 파일 인풋으로 사진 받고 받아지면 시퀀스 이동
- (모바일) 카메라버튼 선택의 경우 ***navigator.mediaDevices.getUserMedia(options) >연구필요 pwa 기본설정 필요***
- options = { audio: false, video: { facingMode: { exact: "environment" } } } >연구필요
- (이게 될까? 된다면 다음으로)
- 7번이 성공한다면 촬영된 사진을 가지고 다음 시퀀스로 이동
- 배경에는 촬영하거나 선택된 사진이 있고
- 상단에 텍스트 / 필터 / 스티커???(미정) 이런 버튼들이 있음
- 필터는 테일윈드에 아주 잘되어 있으므로 누를때마다 순차적으로 사진에 필터 적용하면되고
- 텍스트는 오버레이 <div>를 하나 만들어놓고
- 기본 중앙에서 input type=”text” 해서 인풋 받고
- absolute fixed 주고 드래그 되도록... (이것은 라이브러리 찾아볼것) dnd
- 아무튼 드래그가 되어서, onMouseLeave?? 시에 pageX, pageY??? 이게 맞나
- 암튼 그 값을 찾아서 overlay 칼럼에 jsonb 로 저장... { overlay : [ { text : “값", position: [ x, y] }, {}, ... ]}
- 여기까지 되면 성공
- (스토리보여주는 페이지에서는 저것을 이용해 역시 overlay
- 내에 map 돌려서 보여주면 될...까...?)

### 인증/인가 의사코드

**Comment by slsl2** [comment]
### 홈 의사코드

**1. 배너**
trips 테이블에서 get
trips.trip_master ||  trips.trip_members 에 있는 id와 유저 id가 일치하는 trip만 필터링
start_data를 기준으로 저장
0번째 인덱스만 불러와서 배너에 노출
(새로 추가된 테이블 기준으로 수정)

**2. 추천 인기 버디즈**
온도 기준으로 정렬
buddies 테이블에서 get
메인페이지에선 몇개만 가져오기
전체보기 누르면 /buddies 페이지로 이동

**3. 인기 스토리**
(이 부분은 은님에게 컴퍼넌트 맡김)

**4. 지금 모집중인 여정**
trips 테이블에서 isValidated 만 필터링해서 get
메인페이지에선 몇개만 가져오기
전체보기 누르면 /trips 페이지로 이동



### chat 의사코드

**1. 채팅 생성:**
trip 모집 생성 시 바로 contracts 테이블에 isLeader=true인 row 생성

**2. contract 추가 및 채팅방 입장:**
채팅하기 = isValidate=true 해당되어야만 채팅창 접근 가능
채팅하기 클릭 -> master에게 요청 알림 -> 거절or수락 창
거절 시 거절됐다는 알림
수락 시 수락됐다는 알림 (클릭하면 채팅으로 이동)
수락 시 participant 테이블에 row 생성
master의 경우: 수락하기 -> 채팅방 이동

**3. 채팅 목록:**
contracts 테이블에서 user_id와 member_id? 비교 돌려서 해당되는 채팅만 가져옴
상태에 저장, 목록 렌더링
클릭하면 chat/[id]로 이동

**4. 메시지 전송:**
해당 chat과 연결된 messages 리스트를 상태에 저장
메시지 도착할 때마다 메시지 리스트 업데이트
input 창에 입력한 message_content와 message_sender_id 저장
전송 누르면 messages 테이블에 새로운 메시지 삽입
메시지 보내면 notification 테이블에 type="message"로 새 row 추가

**5. 이미지 전송 로직**
- createObjectUrl
이미지를 화면에서 볼 수 있도록 임시 url 생성
- supabase storage에 저장 (webp로)
- getPublicUrl 사용해서 Image 컴퍼넌트로 뿌리기

**Comment by Kyara0324** [comment]
### 1. 필요한 라이브러리 불러오기
- 1.1. 토스 결제 API와 필요한 모듈을 불러온다.
- 1.2. 결제 API 키를 설정한다.

### 2. 결제 페이지 구성
- 2.1. HTML 구조를 만든다.
    -  2.1.1. 상품 정보 표시
    -  2.1.2. 결제 버튼
- 2.2. 결제 버튼에 클릭 이벤트 리스너를 추가한다.

### 3. 결제 요청 및 처리
- 3.1. 결제 버튼 클릭 시 실행할 함수를 정의한다.
    -  3.1.1. 결제 정보를 구성한다 (상품명, 가격 등).
    -  3.1.2. 토스 결제 API를 호출하여 결제를 요청한다.
    -  3.1.3. 결제 요청에 대한 응답을 처리한다 (성공/실패).

### 4. 결제 완료 후 처리
- 4.1. 결제 성공 시 사용자에게 성공 메시지를 표시한다.
- 4.2. 결제 실패 시 사용자에게 실패 메시지를 표시한다.

**Comment by hyunjeongkwak** [comment]
### 오늘 한 일

의사코드 공유, 디자이너님과 와이어 프레임 확정 및 디자인에 관한 의견 공유, DB테이블 설계, 페이지 및 컴포넌트 분배 확정

#### 의사코드
  - **검색**
  - 1. 키워드 검색 : 키워드 입력 - 키워드 제공되었으면 필터링된 검색 결과 페이지로 이동(자동 완성, 오타 수정)
    - 1. 가이드 문서 배포 : Jenkins를 통해 guide문서를 guide 서버에 배포
    - 2. 가이드 문서 parsing : Docsearch를 이용하여 가이드 문서를 parsing
    - 3. algolia service로 data 인덱싱: parsing 한 데이터는 Docsearch에 의해 자동으로 algolia service에 인덱싱된다고 함
    - 4. 가이드 페이지에서 검색 : 검색 시 docsearch.js를 활용하여 algolia api를 호출
    - 5. api호출에 대한 response: 검색 결과를 받아 화면에 노출

  - 2. 지역 및 국가 검색 : 위치 검색 모달 -  국내/해외 선택 토글 - 하위 요소(지역/대륙 및 국가, 지역) 선택 후 검색 결과 페이지로 이동
  - 3. 날짜 및 기간 검색 : 날짜 선택 모달(달력 라이브러리) - 날짜 선택 후 검색 결과 페이지로 이동
  - 4. 여정 테마 또는 버디즈 성향이 제공된다면
    - 여정 테마 있다면, 여정 테이블에서 선택한 여정 유형 일치 여부 확인, 일치하는 여정 유형 개수 카운트
    - 버디즈 성향이 있다면, 여정 테이블에서 선호 버디즈 유형 일치 여부 확인, 일치하는 버디즈 유형 개수 카운트
  - 5. 검색결과페이지 정렬
    - 여정 검색 결과 : 일치하는 칩스 카운트가 높을수록 위에 뜸
    - 성향 맞는 버디즈 추천 : 온보딩에 작성된 본인의 성향과 칩스 카운트 높은 순 몇 명만 추천

  - **검색결과페이지**
  - 1. 여정 검색 결과 : 여정만(만약 없어도 추천 컨텐츠는 렌더링)
  - 2. 추천컨텐츠 : 온보딩 시 입력한 선호 여행 테마 기반으로 여정 추천(임박한 여정 혹은 남은 자리가 적은 여정 순으로 정렬)
